File: src/main/java/com/oracle/tool/Orchestrator.java
package com.oracle.tool;

import java.io.IOException;
import java.io.InputStream;
import java.io.Reader;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.SQLException;
import java.sql.Types;
import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * High-level execution flow for Download and Upload modes.
 */
public class Orchestrator {
  private static final Logger logger = LoggerFactory.getLogger(Orchestrator.class);

  private final InputManager inputManager;
  private final OracleConnector dbConnector;
  private final CLOBProcessor clobProcessor;
  private final FSManager fsManager;

  public Orchestrator(InputManager inputManager,
      OracleConnector dbConnector,
      CLOBProcessor clobProcessor,
      FSManager fsManager) {
    this.inputManager = inputManager;
    this.dbConnector = dbConnector;
    this.clobProcessor = clobProcessor;
    this.fsManager = fsManager;
  }

  /**
   * Orchestrates UC-1: Download CLOBs.
   *
   * @param csvPath   Path to the CSV file containing IDs.
   * @param outputDir Directory where CLOBs will be saved.
   * @param dbConfig  Database configuration.
   * @throws IOException  If an I/O error occurs.
   * @throws SQLException If a database access error occurs.
   */
  public void downloadMode(Path csvPath, Path outputDir, DBConfig dbConfig)
      throws IOException, SQLException {
    downloadMode(csvPath, outputDir, dbConfig, null);
  }

  /**
   * Orchestrates UC-1: Download CLOBs with progress reporting.
   *
   * @param csvPath   Path to the CSV file containing IDs (optional if idQuery is provided).
   * @param outputDir Directory where CLOBs will be saved.
   * @param dbConfig  Database configuration.
   * @param reporter  Progress reporter (optional).
   * @throws IOException  If an I/O error occurs.
   * @throws SQLException If a database access error occurs.
   */
  public void downloadMode(Path csvPath, Path outputDir, DBConfig dbConfig,
      ProgressReporter reporter)
      throws IOException, SQLException {
    dbConnector.connect(dbConfig);
    try {
      List<String> ids;
      if (dbConfig.idQuery() != null && !dbConfig.idQuery().isEmpty()) {
        logger.info("Fetching IDs using query: {}", dbConfig.idQuery());
        ids = dbConnector.fetchIds(dbConfig.idQuery());
      } else if (csvPath != null) {
        ids = inputManager.loadIds(csvPath);
      } else {
        throw new IllegalArgumentException("Either csvPath or idQuery must be provided.");
      }

      if (ids.isEmpty()) {
        logger.info("No IDs found.");
        return;
      }

      fsManager.ensureDirectory(outputDir);

      Stream<LobRecord> clobStream;
      if (ids.size() < 1000) {
        logger.info("Using IN clause strategy for {} IDs", ids.size());
        if (reporter != null) {
          reporter.setTotal(ids.size());
        }
        clobStream = dbConnector.fetchClobsIn(ids);
      } else {
        logger.info("Using GTT Join strategy for {} IDs", ids.size());
        if (reporter != null) {
          reporter.setTotal(ids.size());
        }
        dbConnector.createGtt(ids);
        clobStream = dbConnector.fetchClobsJoin();
      }

      try (clobStream) {
        Iterable<LobRecord> iterable = clobStream::iterator;
        for (LobRecord record : iterable) {
          String fileName = record.filename();
          if (fileName == null || fileName.isEmpty()) {
            fileName = record.id() + ".txt";
          }
          Path targetPath = outputDir.resolve(fileName);
          clobProcessor.streamToFile(record.lob(), targetPath);
          if (reporter != null) {
            reporter.update(1);
          }
        }
      } catch (RuntimeException ex) {
        if (ex.getCause() instanceof SQLException sqlException) {
          throw sqlException;
        }
        throw ex;
      }
    } finally {
      if (reporter != null) {
        reporter.finish();
      }
      dbConnector.close();
    }
  }

  /**
   * Orchestrates UC-2: Upload CLOBs.
   *
   * @param csvPath  Path to the CSV file containing IDs.
   * @param inputDir Directory containing files to upload.
   * @param dbConfig Database configuration.
   * @throws IOException  If an I/O error occurs.
   * @throws SQLException If a database access error occurs.
   */
  public void uploadMode(Path csvPath, Path inputDir, DBConfig dbConfig)
      throws IOException, SQLException {
    uploadMode(csvPath, inputDir, dbConfig, false, 100);
  }

  /**
   * Orchestrates UC-2: Upload CLOBs with optional regex matching and batch size.
   *
   * @param csvPath     Path to the CSV file containing IDs (optional if idQuery is provided).
   * @param inputDir    Directory containing files to upload.
   * @param dbConfig    Database configuration.
   * @param idAsRegex   Whether to treat IDs as regex patterns.
   * @param batchSize   Batch size for periodic commits.
   * @throws IOException  If an I/O error occurs.
   * @throws SQLException If a database access error occurs.
   */
  public void uploadMode(Path csvPath, Path inputDir, DBConfig dbConfig, boolean idAsRegex,
      int batchSize)
      throws IOException, SQLException {
    dbConnector.connect(dbConfig);
    try {
      List<String> patternsOrIds;
      if (dbConfig.idQuery() != null && !dbConfig.idQuery().isEmpty()) {
        logger.info("Fetching IDs using query: {}", dbConfig.idQuery());
        patternsOrIds = dbConnector.fetchIds(dbConfig.idQuery());
      } else if (csvPath != null) {
        patternsOrIds = inputManager.loadIds(csvPath);
      } else {
        throw new IllegalArgumentException("Either csvPath or idQuery must be provided.");
      }

      if (patternsOrIds.isEmpty()) {
        logger.info("No IDs found.");
        return;
      }

      OracleConnector.LobMetadata metadata = dbConnector.getLobColumnMetadata();
      boolean isBinary = (metadata.type() == Types.BLOB);
      boolean isXml = metadata.typeName() != null && metadata.typeName().toUpperCase().contains("XMLTYPE");
      int uploadAttempted = 0;
      int uploadSuccess = 0;

      List<AutoCloseable> resources = new ArrayList<>();
      List<String> currentBatchIds = new ArrayList<>();
      List<Path> currentBatchPaths = new ArrayList<>();

      try {
        if (idAsRegex) {
          List<Pattern> compiledPatterns = new ArrayList<>();
          for (String p : patternsOrIds) {
            try {
              compiledPatterns.add(Pattern.compile(p));
            } catch (Exception e) {
              logger.error("Invalid regex pattern '{}': {}", p, e.getMessage());
            }
          }

          try (Stream<Path> paths = Files.list(inputDir)) {
            Iterable<Path> filePaths = paths.filter(Files::isRegularFile)::iterator;
            for (Path filePath : filePaths) {
              String filename = filePath.getFileName().toString();
              for (Pattern pattern : compiledPatterns) {
                Matcher matcher = pattern.matcher(filename);
                if (matcher.find()) {
                  String dbId = matcher.groupCount() > 0 ? matcher.group(1) : matcher.group(0);
                  logger.info("Matched file {} with pattern {} -> ID: {}",
                      filename, pattern.pattern(), dbId);

                  if (isXml) {
                    CLOBProcessor.ValidationResult result = clobProcessor.validateXml(filePath);
                    if (!result.valid()) {
                      logger.error("Invalid XML content in file {}. Skipping upload for ID {}. Error: {}",
                          filePath.getFileName(), dbId, result.errorMessage());
                      continue;
                    }
                  }

                  addFileToBatch(dbId, filePath, isBinary, resources, currentBatchIds, currentBatchPaths);
                  uploadAttempted++;

                  if (uploadAttempted % batchSize == 0) {
                    uploadSuccess += executeBatchAndClose(resources, currentBatchIds, currentBatchPaths);
                  }
                  break;
                }
              }
            }
          }
        } else {
          for (String idVal : patternsOrIds) {
            Path filePath = inputDir.resolve(idVal + ".txt");
            if (!Files.exists(filePath)) {
              filePath = inputDir.resolve(idVal);
            }
            if (!Files.exists(filePath)) {
              try (Stream<Path> matches = Files.list(inputDir)) {
                filePath = matches
                    .filter(p -> p.getFileName().toString().startsWith(idVal + "."))
                    .findFirst()
                    .orElse(filePath);
              }
            }

            if (Files.exists(filePath)) {
              if (isXml) {
                CLOBProcessor.ValidationResult result = clobProcessor.validateXml(filePath);
                if (!result.valid()) {
                  logger.error("Invalid XML content in file {}. Skipping upload for ID {}. Error: {}",
                      filePath.getFileName(), idVal, result.errorMessage());
                  continue;
                }
              }

              logger.info("Uploading file {} for ID {}", filePath.getFileName(), idVal);
              addFileToBatch(idVal, filePath, isBinary, resources, currentBatchIds, currentBatchPaths);
              uploadAttempted++;

              if (uploadAttempted % batchSize == 0) {
                uploadSuccess += executeBatchAndClose(resources, currentBatchIds, currentBatchPaths);
              }
            } else {
              logger.warn("File not found for ID {}: {}", idVal, filePath);
            }
          }
        }
        uploadSuccess += executeBatchAndClose(resources, currentBatchIds, currentBatchPaths);
        logger.info("Total files attempted: {}, Successfully updated: {}",
            uploadAttempted, uploadSuccess);
      } finally {
        for (AutoCloseable res : resources) {
          try {
            res.close();
          } catch (Exception e) {
            logger.error("Error closing resource: {}", e.getMessage());
          }
        }
      }
    } finally {
      dbConnector.close();
    }
  }

  private void addFileToBatch(String id, Path filePath, boolean isBinary,
      List<AutoCloseable> resources, List<String> currentBatchIds, List<Path> currentBatchPaths)
      throws IOException, SQLException {
    if (isBinary) {
      InputStream is = clobProcessor.openFileAsStream(filePath);
      resources.add(is);
      dbConnector.addUpdateToBatch(id, is);
    } else {
      Reader reader = clobProcessor.openFile(filePath);
      resources.add(reader);
      dbConnector.addUpdateToBatch(id, reader);
    }
    currentBatchIds.add(id);
    currentBatchPaths.add(filePath);
  }

  private int executeBatchAndClose(List<AutoCloseable> resources, List<String> currentBatchIds,
      List<Path> currentBatchPaths)
      throws SQLException {
    if (currentBatchIds.isEmpty()) {
      return 0;
    }

    int successCount = 0;
    try {
      int[] results = dbConnector.executeUpdateBatch();
      for (int i = 0; i < results.length; i++) {
        if (results[i] > 0 || results[i] == -2) { // -2 is SUCCESS_NO_INFO
          successCount++;
        } else if (results[i] == 0) {
          logger.warn("No rows updated for ID {}. Record may not exist.", currentBatchIds.get(i));
        }
      }
      dbConnector.commit();
    } catch (SQLException ex) {
      logger.error("Batch update failed. Data preview for failed batch:");
      for (Path path : currentBatchPaths) {
        try {
          List<String> lines = Files.readAllLines(path);
          logger.error("File: {} - First 3 lines:", path.getFileName());
          lines.stream().limit(3).forEach(line -> logger.error("  {}", line));
        } catch (IOException e) {
          logger.error("Could not read file {} for logging: {}", path, e.getMessage());
        }
      }
      throw ex;
    } finally {
      for (AutoCloseable res : resources) {
        try {
          res.close();
        } catch (Exception e) {
          logger.error("Error closing resource: {}", e.getMessage());
        }
      }
      resources.clear();
      currentBatchIds.clear();
      currentBatchPaths.clear();
    }

    return successCount;
  }
}

File: src/orchestrator.py
from pathlib import Path
from src.input_manager import InputManager
from src.oracle_connector import OracleConnector, DBConfig
from src.clob_processor import CLOBProcessor
import logging
from src.fs_manager import FSManager
from typing import Optional

logger = logging.getLogger(__name__)

class ProgressReporter:
    """Base class/interface for progress reporting."""
    def set_total(self, total: int):
        pass

    def update(self, n: int):
        pass

    def finish(self):
        pass

class Orchestrator:
    """High-level execution flow for Download and Upload modes."""

    def __init__(self,
                 input_manager: InputManager,
                 db_connector: OracleConnector,
                 clob_processor: CLOBProcessor,
                 fs_manager: FSManager):
        self.input_manager = input_manager
        self.db_connector = db_connector
        self.clob_processor = clob_processor
        self.fs_manager = fs_manager

    def download_mode(self, csv_path: Optional[Path], output_dir: Path, db_config: DBConfig,
                      reporter: Optional[ProgressReporter] = None):
        """Orchestrates UC-1."""
        self.db_connector.connect(db_config)
        try:
            if db_config.id_query:
                logger.info(f"Fetching IDs using query: {db_config.id_query}")
                ids = self.db_connector.fetch_ids(db_config.id_query)
            elif csv_path:
                ids = self.input_manager.load_ids(csv_path)
            else:
                raise ValueError("Either csv_path or id_query must be provided.")

            if not ids:
                logger.info("No IDs found.")
                return

            self.fs_manager.ensure_directory(output_dir)

            if len(ids) < 1000:
                logger.info(f"Using IN clause strategy for {len(ids)} IDs")
                if reporter:
                    reporter.set_total(len(ids))
                clob_iterator = self.db_connector.fetch_clobs_in(ids)
            else:
                logger.info(f"Using GTT Join strategy for {len(ids)} IDs")
                if reporter:
                    reporter.set_total(len(ids))
                self.db_connector.create_gtt(ids)
                clob_iterator = self.db_connector.fetch_clobs_join()

            for row in clob_iterator:
                id_val, clob_lob = row[0], row[1]
                filename = row[2] if len(row) > 2 else None

                if filename:
                    target_path = output_dir / filename
                else:
                    target_path = output_dir / f"{id_val}.txt"

                self.clob_processor.stream_to_file(clob_lob, target_path)
                if reporter:
                    reporter.update(1)
        finally:
            if reporter:
                reporter.finish()
            self.db_connector.close()

    def upload_mode(self, csv_path: Optional[Path], input_dir: Path, db_config: DBConfig,
                    id_as_regex: bool = False, batch_size: int = 100):
        """Orchestrates UC-2."""
        import oracledb
        self.db_connector.connect(db_config)
        try:
            if db_config.id_query:
                logger.info(f"Fetching IDs using query: {db_config.id_query}")
                patterns_or_ids = self.db_connector.fetch_ids(db_config.id_query)
            elif csv_path:
                patterns_or_ids = self.input_manager.load_ids(csv_path)
            else:
                raise ValueError("Either csv_path or id_query must be provided.")

            if not patterns_or_ids:
                logger.info("No IDs found.")
                return

            col_type = self.db_connector.get_lob_column_type()
            is_binary = (col_type == oracledb.DB_TYPE_BLOB)
            mode = 'rb' if is_binary else 'r'

            upload_attempted = 0
            upload_success = 0

            def _perform_update(db_id_val, f_obj):
                nonlocal upload_success
                affected = self.db_connector.update_lob(db_id_val, f_obj)
                if affected > 0:
                    upload_success += 1
                else:
                    logger.warning(f"No rows updated for ID {db_id_val}. Record may not exist.")

            if id_as_regex:
                import re
                compiled_patterns = []
                for p in patterns_or_ids:
                    try:
                        compiled_patterns.append(re.compile(p))
                    except re.error as e:
                        logger.error(f"Invalid regex pattern '{p}': {e}")

                for file_path in input_dir.iterdir():
                    if not file_path.is_file():
                        continue
                    filename = file_path.name
                    for cp in compiled_patterns:
                        match = cp.search(filename)
                        if match:
                            db_id = match.group(1) if match.groups() else match.group(0)
                            logger.info(f"Matched file {filename} with pattern {cp.pattern} -> ID: {db_id}")
                            with self.clob_processor.open_file(file_path, mode=mode) as f:
                                _perform_update(db_id, f)
                            upload_attempted += 1
                            if upload_attempted % batch_size == 0:
                                self.db_connector.commit()
                            break
            else:
                # Try with .txt, then without extension if not found
                for id_val in patterns_or_ids:
                    file_path = input_dir / f"{id_val}.txt"
                    if not file_path.exists():
                        file_path = input_dir / id_val

                    if not file_path.exists():
                        # Try globbing
                        matches = list(input_dir.glob(f"{id_val}.*"))
                        if matches:
                            file_path = matches[0]

                    if file_path.exists():
                        logger.info(f"Uploading file {file_path.name} for ID {id_val}")
                        with self.clob_processor.open_file(file_path, mode=mode) as f:
                            _perform_update(id_val, f)
                        upload_attempted += 1
                        if upload_attempted % batch_size == 0:
                            self.db_connector.commit()
                    else:
                        logger.warning(f"File not found for ID {id_val} in {input_dir}")

            self.db_connector.commit()
            logger.info(f"Total files attempted: {upload_attempted}, Successfully updated: {upload_success}")
        finally:
            self.db_connector.close()
